plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/simulate/ParsedSimulateRequest.java

```diff
@@ -0,0 +1,101 @@
+import org.elasticsearch.ingest.processor.ConfigurationUtils;
+import org.elasticsearch.plugin.ingest.PipelineStore;
+
+import java.io.IOException;
+import java.util.*;
+
+public class ParsedSimulateRequest {
+    private final List<Data> documents;
+    private final Pipeline pipeline;
+    private final boolean verbose;
+
+    ParsedSimulateRequest(Pipeline pipeline, List<Data> documents, boolean verbose) {
+        this.pipeline = pipeline;
+        this.documents = Collections.unmodifiableList(documents);
+        this.verbose = verbose;
+    }
+
+    public Pipeline getPipeline() {
+        return pipeline;
+    }
+
+    public List<Data> getDocuments() {
+        return documents;
+    }
+
+    public boolean isVerbose() {
+        return verbose;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ParsedSimulateRequest that = (ParsedSimulateRequest) o;
+        return Objects.equals(verbose, that.verbose) &&
+                Objects.equals(documents, that.documents) &&
+                Objects.equals(pipeline, that.pipeline);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(documents, pipeline, verbose);
+    }
+
+    public static class Parser {
+        private static final Pipeline.Factory PIPELINE_FACTORY = new Pipeline.Factory();
+        public static final String SIMULATED_PIPELINE_ID = "_simulate_pipeline";
+
+        public ParsedSimulateRequest parse(String pipelineId, Map<String, Object> config, boolean verbose, PipelineStore pipelineStore) throws IOException {
+            Pipeline pipeline;
+            // if pipeline `id` passed to request, fetch pipeline from store.
+            if (pipelineId != null) {
+                pipeline = pipelineStore.get(pipelineId);
+            } else {
+                Map<String, Object> pipelineConfig = ConfigurationUtils.readOptionalMap(config, "pipeline");
+                pipeline = PIPELINE_FACTORY.create(SIMULATED_PIPELINE_ID, pipelineConfig, pipelineStore.getProcessorFactoryRegistry());
+            }
+
+            List<Map<String, Object>> docs = ConfigurationUtils.readList(config, "docs");
+
+            List<Data> dataList = new ArrayList<>();
+
+            for (int i = 0; i < docs.size(); i++) {
+                Map<String, Object> dataMap = docs.get(i);
+                Map<String, Object> document = ConfigurationUtils.readOptionalMap(dataMap, "_source");
+                if (document == null) {
+                    document = Collections.emptyMap();
+                }
+                Data data = new Data(ConfigurationUtils.readOptionalStringProperty(dataMap, "_index"),
+                        ConfigurationUtils.readOptionalStringProperty(dataMap, "_type"),
+                        ConfigurationUtils.readOptionalStringProperty(dataMap, "_id"),
+                        document);
+                dataList.add(data);
```

Comment:
I see your point I just don't want to create ambiguity with this leniency. At the end of the day source is mandatory for processors that access it, same for metadata, but then what do we do when we need it and it's not provided? I can imagine processors will be utterly broken as they assume the required data to be there, that is why I think both metadata and source should be mandatory and we should enforce this.